pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go

```diff
@@ -1131,11 +1133,7 @@ func (up *upgradePlan) Add(steps []testStep) {
 	up.sequentialStep.steps = append(up.sequentialStep.steps, steps...)
 }
 
-// mapSingleSteps iterates over every step in the test plan and calls
-// the given function `f` for every `singleStep` (i.e., every step
-// that actually performs an action). The function should return a
-// list of testSteps that replace the given step in the plan.
-func (plan *TestPlan) mapSingleSteps(f func(*singleStep, bool) []testStep) {
+func (plan *TestPlan) walkSteps(steps []testStep, f func(*singleStep, bool) []testStep) []testStep {
```

Comment:
Just refactoring for `iterateSingleSteps`, below, so that we can reuse most of the code; no changes other than lifting out the helper funcs.

Reply:
Good catch! I had a nagging suspicion it wasn't pure, but I missed that it was actually mutating steps. I'll see what can be done with minimal refactoring.